crates/bevy_ecs/src/storage/blob_vec.rs

```diff
@@ -353,6 +380,9 @@ impl Drop for BlobVec {
 fn array_layout(layout: &Layout, n: usize) -> Option<Layout> {
     let (array_layout, offset) = repeat_layout(layout, n)?;
     debug_assert_eq!(layout.size(), offset);
+
+    //Note: NEEDED for batching.  This is the layout of the array itself, not the layout of its elements.
+    let array_layout = array_layout.align_to(batch::MAX_SIMD_ALIGNMENT).unwrap();
```

Comment:
Sorry I can see where I wasn't clear with this.  `stride == size` definitely still holds -- this statement just aligns the *start* of the array, if that makes sense (the chunk of memory for the array *itself* is aligned to `MAX_SIMD_ALIGNMENT`).  The element layout remains exactly the same as it was before.  This was needed to guarantee that a batch starting at index 0 would be aligned.

For interested readers, typically SIMD code operates in three "phases", a prologue, a "vector region", and an epilogue.
The prologue and epilogue are just "scalar code paths", i.e., your typical for loops.  The prologue's purpose is to process elements up until it is possible to execute vector instructions, which generally require a certain alignment to work.  However, the vector region can only execute in batches of size `N`, which means that if there are any remaining items to do that can't fit in a batch of size `N`, they have to be processed serially.


As an example, consider an array of `[f32; 128]`, which is aligned to a boundary of 4 (say, address `100`).  Suppose you had a special vector instruction set that could operate on batches of 8 `f32`s at a time, the catch being, the loads have to be aligned to 32 as well (sizeof `f32` being `4`).  Then, because of the alignment of the array, you would only be able to batch from index 8 onwards.  Then you would of course need a prologue to handle that "slow path".  However, if the array started at address `96`, you could see that no prologue would be necessary.

This line here eliminates the prologue case from Bevy by guaranteeing that batch 0 is aligned to the correct SIMD alignment in all cases, because `MAX_SIMD_ALIGNMENT` is the greatest alignment required of any instruction on the given system (including vector instructions).  Does this make sense?  Thank you!

Reply:
stride == size does not hold for the layout returned by this fn, `align(MAX_SIMD_ALIGNMENT) size(2)` for example isnt stride == size but could be potentially returned from this fn

Reply:
Yes that's exactly what I was after -- this `array_layout` is what is ultimately used to allocate memory for the entire array *after* the stride has already been taken into account.  It is derived from the `item_layout` using the `repeat_layout` function.  The flow (as I understand it) works as follows:

1 - The `item_layout` of `T` is passed in via the `BlobVec` constructor and held. (stride matters here, as you rightly point out!)
2 -  The `array_layout` is generated from `repeat_layout(layout, n)` (stride is taken into account by repeat_layout -- this gives us a Layout that can hold N `item_layout`s [size == stride] and aligned to alignment of item_layout)
3 -  The "SIMD aligned" `array_layout` is obtained by taking that `array_layout` and then calling `align_to` on it with `MAX_SIMD_ALIGNMENT` (stride has already been taken into account -- we are simply telling the allocator "take this Layout and ensure that the alignment requirement is `MAX_SIMD_ALIGNMENT`.  I.e. this is now a Layout that can hold N `item_layout`s [size == stride] except the the chunk of memory will now be aligned to MAX_SIMD_ALIGNMENT)
4 - It is ultimately (3) that is passed into the allocator

Note that the `item_layout` is left pristine through all of this to ensure we don't add any extra padding between elements.  It is strictly just the `array_layout` that gets `align_to(MAX_SIMD_ALIGNMENT)`.
The actual elements of the array are still laid out like `repeat_layout` would do it.

This is the best of my understanding of that at the moment.  If the code called `align_to` on the `item_layout` itself, then `stride != size` for sure.  But since it's being called on `array_layout` *after* `repeat_layout` has already done its job, the stride should already have been taken into account.  Is my understanding right here?

(Apologies for the ASCII art, I'm a visual thinker :D)

~~~

1 - item_layout provided to constructor
2 - align(item_layout):        | item_layout item_layout item_layout item_layout ... (N times) | (repeat_layout)
3 - align(MAX_SIMD_ALIGNMENT): | item_layout item_layout item_layout item_layout ... (N times) | (array_layout)

~~~

Reply:
A rust playground link illustrating it a little further: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=099b1168135ed7ed189dd3bad76477c1

Reply:
Since it's been a while, I'm just posting an [updated gist link with code embedded](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=use%20core%3A%3Aalloc%3A%3ALayout%3B%0A%0Aconst%20MAX_SIMD_ALIGNMENT%3A%20usize%20%3D%2064%3B%0A%0A%2F%2F%2F%20From%20%3Chttps%3A%2F%2Fdoc.rust-lang.org%2Fbeta%2Fsrc%2Fcore%2Falloc%2Flayout.rs.html%3E%0Afn%20simd_array_layout(layout%3A%20%26Layout%2C%20n%3A%20usize)%20-%3E%20Option%3CLayout%3E%20%7B%0A%20%20%20%20let%20(array_layout%2C%20offset)%20%3D%20repeat_layout(layout%2C%20n)%3F%3B%0A%20%20%20%20debug_assert_eq!(layout.size()%2C%20offset)%3B%0A%0A%20%20%20%20%2F%2FNote%3A%20NEEDED%20for%20batching.%20%20This%20is%20the%20layout%20of%20the%20array%20itself%2C%20not%20the%20layout%20of%20its%20elements.%0A%20%20%20%20let%20array_layout%20%3D%20array_layout.align_to(MAX_SIMD_ALIGNMENT).unwrap()%3B%0A%20%20%20%20Some(array_layout)%0A%7D%0A%0A%2F%2F%2F%20From%20%3Chttps%3A%2F%2Fdoc.rust-lang.org%2Fbeta%2Fsrc%2Fcore%2Falloc%2Flayout.rs.html%3E%0A%2F%2F%20Original%20unmodified%20array_layout()%20function%0Afn%20original_array_layout(layout%3A%20%26Layout%2C%20n%3A%20usize)%20-%3E%20Option%3CLayout%3E%20%7B%0A%20%20%20%20let%20(array_layout%2C%20offset)%20%3D%20repeat_layout(layout%2C%20n)%3F%3B%0A%20%20%20%20debug_assert_eq!(layout.size()%2C%20offset)%3B%0A%0A%20%20%20%20Some(array_layout)%0A%7D%0A%0A%2F%2F%20TODO%3A%20replace%20with%20%60Layout%3A%3Arepeat%60%20if%2Fwhen%20it%20stabilizes%0A%2F%2F%2F%20From%20%3Chttps%3A%2F%2Fdoc.rust-lang.org%2Fbeta%2Fsrc%2Fcore%2Falloc%2Flayout.rs.html%3E%0Afn%20repeat_layout(layout%3A%20%26Layout%2C%20n%3A%20usize)%20-%3E%20Option%3C(Layout%2C%20usize)%3E%20%7B%0A%20%20%20%20%2F%2F%20This%20cannot%20overflow.%20Quoting%20from%20the%20invariant%20of%20Layout%3A%0A%20%20%20%20%2F%2F%20%3E%20%60size%60%2C%20when%20rounded%20up%20to%20the%20nearest%20multiple%20of%20%60align%60%2C%0A%20%20%20%20%2F%2F%20%3E%20must%20not%20overflow%20(i.e.%2C%20the%20rounded%20value%20must%20be%20less%20than%0A%20%20%20%20%2F%2F%20%3E%20%60usize%3A%3AMAX%60)%0A%20%20%20%20let%20padded_size%20%3D%20layout.size()%20%2B%20padding_needed_for(layout%2C%20layout.align())%3B%0A%20%20%20%20let%20alloc_size%20%3D%20padded_size.checked_mul(n)%3F%3B%0A%0A%20%20%20%20%2F%2F%20SAFETY%3A%20self.align%20is%20already%20known%20to%20be%20valid%20and%20alloc_size%20has%20been%0A%20%20%20%20%2F%2F%20padded%20already.%0A%20%20%20%20unsafe%20%7B%0A%20%20%20%20%20%20%20%20Some((%0A%20%20%20%20%20%20%20%20%20%20%20%20Layout%3A%3Afrom_size_align_unchecked(alloc_size%2C%20layout.align())%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20padded_size%2C%0A%20%20%20%20%20%20%20%20))%0A%20%20%20%20%7D%0A%7D%0A%0A%2F%2F%2F%20From%20%3Chttps%3A%2F%2Fdoc.rust-lang.org%2Fbeta%2Fsrc%2Fcore%2Falloc%2Flayout.rs.html%3E%0Aconst%20fn%20padding_needed_for(layout%3A%20%26Layout%2C%20align%3A%20usize)%20-%3E%20usize%20%7B%0A%20%20%20%20let%20len%20%3D%20layout.size()%3B%0A%0A%20%20%20%20%2F%2F%20Rounded%20up%20value%20is%3A%0A%20%20%20%20%2F%2F%20%20%20len_rounded_up%20%3D%20(len%20%2B%20align%20-%201)%20%26%20!(align%20-%201)%3B%0A%20%20%20%20%2F%2F%20and%20then%20we%20return%20the%20padding%20difference%3A%20%60len_rounded_up%20-%20len%60.%0A%20%20%20%20%2F%2F%0A%20%20%20%20%2F%2F%20We%20use%20modular%20arithmetic%20throughout%3A%0A%20%20%20%20%2F%2F%0A%20%20%20%20%2F%2F%201.%20align%20is%20guaranteed%20to%20be%20%3E%200%2C%20so%20align%20-%201%20is%20always%0A%20%20%20%20%2F%2F%20%20%20%20valid.%0A%20%20%20%20%2F%2F%0A%20%20%20%20%2F%2F%202.%20%60len%20%2B%20align%20-%201%60%20can%20overflow%20by%20at%20most%20%60align%20-%201%60%2C%0A%20%20%20%20%2F%2F%20%20%20%20so%20the%20%26-mask%20with%20%60!(align%20-%201)%60%20will%20ensure%20that%20in%20the%0A%20%20%20%20%2F%2F%20%20%20%20case%20of%20overflow%2C%20%60len_rounded_up%60%20will%20itself%20be%200.%0A%20%20%20%20%2F%2F%20%20%20%20Thus%20the%20returned%20padding%2C%20when%20added%20to%20%60len%60%2C%20yields%200%2C%0A%20%20%20%20%2F%2F%20%20%20%20which%20trivially%20satisfies%20the%20alignment%20%60align%60.%0A%20%20%20%20%2F%2F%0A%20%20%20%20%2F%2F%20(Of%20course%2C%20attempts%20to%20allocate%20blocks%20of%20memory%20whose%0A%20%20%20%20%2F%2F%20size%20and%20padding%20overflow%20in%20the%20above%20manner%20should%20cause%0A%20%20%20%20%2F%2F%20the%20allocator%20to%20yield%20an%20error%20anyway.)%0A%0A%20%20%20%20let%20len_rounded_up%20%3D%20len.wrapping_add(align).wrapping_sub(1)%20%26%20!align.wrapping_sub(1)%3B%0A%20%20%20%20len_rounded_up.wrapping_sub(len)%0A%7D%0A%0A%2F%2FS%20should%20have%20align_of(i32)%0Astruct%20S%0A%7B%0A%20%20%20%20x%3A%20i32%0A%7D%0A%0Afn%20main()%0A%7B%0A%20%20%20%20let%20s_layout%20%3D%20Layout%3A%3Anew%3A%3A%3CS%3E()%3B%0A%0A%20%20%20%20let%20simd_blob_vec_layout%20%3D%20simd_array_layout(%26s_layout%2C%2053).unwrap()%3B%0A%20%20%20%20%0A%20%20%20%20let%20blob_vec_layout%20%3D%20original_array_layout(%26s_layout%2C%2053).unwrap()%3B%0A%20%20%20%20%0A%20%20%20%20println!(%22simd_blob_vec_layout%3A%20%7B%3A%3F%7D%22%2C%20simd_blob_vec_layout)%3B%0A%20%20%20%20println!(%22blob_vec_layout%3A%20%7B%3A%3F%7D%22%2C%20blob_vec_layout)%3B%0A%20%20%20%20println!(%22simd_blob_vec_layout.size()%20%3D%3D%20simd_blob_vec_layout.size()%3A%20%7B%7D%22%2C%20simd_blob_vec_layout.size()%20%3D%3D%20blob_vec_layout.size())%3B%0A%20%20%20%20%0A%20%20%20%20%2F%2FNote%20that%20the%20sizes%20of%20the%20above%20layouts%20are%20identical!%0A%20%20%20%20%0A%20%20%20%20%0A%7D) for when we come back around to discussing this. in case the original link fails to work.